src/trackedit/internal/au3/au3interaction.cpp

```diff
@@ -117,6 +117,194 @@ muse::Ret Au3Interaction::canPasteClips(const std::vector<TrackId>& dstTracksIds
+            prj->onTrackChanged(DomConverter::track(waveTrack));
+
+            continue;
+        }
+
+        //! NOTE check if clip hovers left side of the clip
+        if (clip->GetPlayStartTime() <= otherClip->GetPlayStartTime()
+            && clip->GetPlayEndTime() <= otherClip->GetPlayEndTime()
+            && clip->GetPlayEndTime() >= otherClip->GetPlayStartTime()) {
+
+            secs_t overlap = clip->GetPlayEndTime() - otherClip->GetPlayStartTime();
+            otherClip->TrimLeft(overlap);
+            prj->onClipChanged(DomConverter::clip(waveTrack, otherClip.get()));
+
+            continue;
+        }
+
+        //! NOTE check if clip hovers right side of the clip
+        if (clip->GetPlayStartTime() >= otherClip->GetPlayStartTime()
+            && clip->GetPlayStartTime() <= otherClip->GetPlayEndTime()
+            && clip->GetPlayEndTime() >= otherClip->GetPlayEndTime()) {
+
+            secs_t overlap = otherClip->GetPlayEndTime() - clip->GetPlayStartTime();
+            otherClip->TrimRight(overlap);
+            prj->onClipChanged(DomConverter::clip(waveTrack, otherClip.get()));
+
+            continue;
+        }
+
+        //! NOTE check if clip boundaries are within other clip
+        if (clip->GetPlayStartTime() >= otherClip->GetPlayStartTime()
+            && clip->GetPlayEndTime() <= otherClip->GetPlayEndTime()) {
+            secs_t otherClipStartTime = otherClip->GetPlayStartTime();
+            secs_t otherClipEndTime = otherClip->GetPlayEndTime();
+
+            auto leftClip = waveTrack->CopyClip(*otherClip, true);
+            waveTrack->InsertInterval(std::move(leftClip), false);
+            prj->onTrackChanged(DomConverter::track(waveTrack));
+
+            secs_t rightClipOverlap = clip->GetPlayEndTime() - otherClip->GetPlayStartTime();
+            otherClip->TrimLeft(rightClipOverlap);
+            prj->onClipChanged(DomConverter::clip(waveTrack, otherClip.get()));
+
+            leftClip->SetPlayStartTime(otherClipStartTime);
+            secs_t leftClipOverlap = otherClipEndTime - clip->GetPlayStartTime();
+            leftClip->TrimRight(leftClipOverlap);
+            prj->onClipChanged(DomConverter::clip(waveTrack, leftClip.get()));
+
+            prj->onTrackChanged(DomConverter::track(waveTrack));
+
+            continue;
+        }
+    }
+
+    return muse::make_ret(muse::Ret::Code::Ok);
+}
+
+muse::Ret Au3Interaction::makeRoomForDataOnTracks(const std::vector<TrackId> &tracksIds, secs_t begin)
+{
+    for (size_t i = 0; i < tracksIds.size(); ++i) {
+        WaveTrack* dstWaveTrack = DomAccessor::findWaveTrack(projectRef(), ::TrackId(tracksIds.at(i)));
+        IF_ASSERT_FAILED(dstWaveTrack) {
+            return make_ret(trackedit::Err::WaveTrackNotFound);
+        }
+
+
+        secs_t insertDuration = clipboard()->trackData(i).track.get()->GetEndTime() - clipboard()->trackData(i).track.get()->GetStartTime();
+        makeRoomForDataOnTrack(tracksIds.at(i), begin, begin + insertDuration);
```

Comment:
If this method returns Ret, then it probably needs to be checked for success and returned further if not successful.